### PR TITLE
virtiofs: Remove feature gate for config volumes

### PR DESCRIPTION
**Pull request by germag**
**What this PR does / why we need it**:
It's currently possible to share config volumes using an unprivileged virtiofs container. So, let's remove the requirement for a feature gate for this type of sharing.

Tested-by: Javier Cano Cano <jcanocan@redhat.com>

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # https://github.com/kubevirt/kubevirt/issues/7735

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Review by xpivarc — CHANGES_REQUESTED** [review]

File: tests/virtiofs/config.go [hunk 1]
```diff
@@ -39,17 +36,18 @@ import (
...
 )
 
-var _ = Describe("[sig-compute] vitiofs config volumes", decorators.SigCompute, func() {
+var _ = Describe("[sig-compute] vitiofs config volumes", Serial, decorators.SigCompute, func() {
 	var virtClient kubecli.KubevirtClient
 
 	BeforeEach(func() {
 		virtClient = kubevirt.Client()
-		checks.SkipTestIfNoFeatureGate(virtconfig.VirtIOFSGate)
+		tests.DisableFeatureGate(virtconfig.VirtIOFSGate)
```
[the remark this answers is not in the document]
Comment (jcanocan) [reply tests/virtiofs/config.go:50]: I have tried to remove that line and add the following one:

```go
BeforeEach(func() {
		virtClient = kubevirt.Client()
		Expect(checks.HasFeature(virtconfig.VirtIOFSGate)).To(BeFalse())
	})
```

The run fails with the following message:

```bash
 Expected
      <bool>: true
  to be false
  In [BeforeEach] at: tests/virtiofs/config.go:52

  Full Stack Trace
    kubevirt.io/kubevirt/tests/virtiofs.glob..func1.1()
    	tests/virtiofs/config.go:52 +0x9e
``` 
Therefore, I think it is required. Looks like by default, in this context, the feature gate is enabled.

[the remark this answers is not in the document]
Comment (xpivarc) [reply tests/virtiofs/config.go:50]: Yes, it is on by default https://github.com/kubevirt/kubevirt/blob/main/tests/testsuite/kubevirtresource.go#L99 . But what is the reason to remove it in test?

[the remark this answers is not in the document]
Comment (germag) [reply tests/virtiofs/config.go:50]: To make sure the FG is not required for config volumes, do you think we don't need to test that in that place?

Maybe I can leave the FG for those tests and add just a single test with all the config volumes and the FG removed, wdyt?

[the remark this answers is not in the document]
Comment (xpivarc) [reply tests/virtiofs/config.go:50]: Yes, I admit I am aiming for optimization as parallel test is much cheaper than serial one.

**Comment by germag** [comment]
v2:
- Keep original functional test to run in parallel
- Add a new test checking if we can share config volumes without the feature gate

File: pkg/virt-controller/services/virtiofs.go [hunk 2]
```diff
@@ -15,6 +15,32 @@ import (
 	"kubevirt.io/kubevirt/pkg/virtiofs"
 )
 
+// RequireAnyVirtiofsPrivilegedContainer Returns true if any of the returns true if any
+// of the filesystems requires a container of virtiofs to run as root
+func RequireAnyVirtiofsPrivilegedContainer(spec *v1.VirtualMachineInstanceSpec) bool {
```
Comment (xpivarc) [inline pkg/virt-controller/services/virtiofs.go:20]: Nit: Can we move this to where we use it?  (Reasoning we are working on dividing packages for different groups. The webhook will be part of sig-api and this function would be part other sig. Meaning the sig-scale would not be able to get the changes in on their own)

**Review by xpivarc — COMMENTED** [review]
/approve
One small nit

**Comment by xpivarc** [comment]
/cc @alicefr 
Can you have a look, please?

File: pkg/virt-controller/services/virtiofs.go (hunk 2 above)
Comment (germag) [reply pkg/virt-controller/services/virtiofs.go:20]: Sure, by doing that, I could also change the functional test (in the last commit) to be just a unit test for the admitter.

But, how can I do it without "leaking" the knowledge how which volumes require root, and which do not require root?
I think this info should, at least, be inside the virtiofsd pakage, would it be enough if I move it to `pkg/virtiofs/virtiofs.go`?

Comment (xpivarc) [reply pkg/virt-controller/services/virtiofs.go:20]: @germag Thanks, now I see the point.
We can expose the `isConfig` or move this as whole. I leave it for you

File: pkg/virt-controller/services/virtiofs.go [hunk 3]
```diff
@@ -15,6 +15,32 @@ import (
 	"kubevirt.io/kubevirt/pkg/virtiofs"
 )
 
+// RequireAnyVirtiofsPrivilegedContainer Returns true if any of the returns true if any
```
Comment (alicefr) [inline pkg/virt-controller/services/virtiofs.go:18]: nit: `RequireAnyVirtiofsPrivilegedContainer Returns true if any of the returns true` -> `RequireAnyVirtiofsPrivilegedContainer returns true`

**Comment by alicefr** [comment]
A part from a small comment for the test for the feature gate, the PR looks good!

/lgtm
/hold
to address the last comments

File: pkg/virt-controller/services/virtiofs.go (hunk 3 above)
Comment (germag) [reply pkg/virt-controller/services/virtiofs.go:18]: was to make sure that people understand that it returns true :-P

File: tests/virtiofs/feature_gate.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+var _ = Describe("[sig-storage] VirtIO-FS feature gate", Serial, decorators.SigStorage, func() {
+	var virtClient kubecli.KubevirtClient
+
+	BeforeEach(func() {
+		virtClient = kubevirt.Client()
+		tests.DisableFeatureGate(virtconfig.VirtIOFSGate)
+	})
+
+	AfterEach(func() {
+		tests.EnableFeatureGate(virtconfig.VirtIOFSGate)
```
Comment (alicefr) [inline tests/virtiofs/feature_gate.go:48]: I know is annoying, but please check if the fg was enabled in the BeforeEach and then enable in the AfterEach only if it was true

**Comment by germag** [comment]
v2:
- Move requireAnyVirtiofsPrivilegedContainer() to the admitter
- Add function to check the required privileges for volumes
- Move feature gates functional tests to their own file
- Having fun getting the order and required blank lines in imports :)

**Pull request merged**